clippy_lints/src/needless_cow.rs

```diff
@@ -0,0 +1,384 @@
+    ///
+    /// let _ = if something {
+    ///     my_owned_string;
+    /// } else if other_thing {
+    ///     String::new();
+    /// } else {
+    ///     String::new();
+    /// };
+    /// ```
+    #[clippy::version = "1.73.0"]
+    pub NEEDLESS_COW,
+    nursery,
+    "usage of `Cow` with only empty strings in `Borrow` variant"
+}
+impl_lint_pass!(NeedlessCow => [NEEDLESS_COW]);
+
+pub struct NeedlessCow;
+
+impl<'tcx> LateLintPass<'tcx> for NeedlessCow {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        _: FnKind<'tcx>,
+        _: &'tcx FnDecl<'tcx>,
+        _: &'tcx Body<'tcx>,
+        _: Span,
+        def_id: LocalDefId,
+    ) {
+        // Probably unnecessary nowadays considering dummy MIR is created for these functions.
+        if fn_has_unsatisfiable_preds(cx, def_id.to_def_id()) {
+            return;
+        }
+
+        let mir = cx.tcx.optimized_mir(def_id);
+        let mut cow_locals_to_assignments = FxIndexMap::default();
+
+        for local in mir
+            .local_decls
+            .iter_enumerated()
+            .filter(|(_, decl)| is_type_diagnostic_item(cx, decl.ty, sym::Cow))
+            .map(|(i, _)| i)
+        {
+            cow_locals_to_assignments.insert(local, vec![]);
+        }
+
+        let mut results = NeedlessCowAnalysis {
+            map: Map::from_filter(cx.tcx, mir, |ty| ty.peel_refs().is_str(), None),
+        }
+        .wrap()
+        .into_engine(cx.tcx, mir)
+        .iterate_to_fixpoint();
+
+        results.visit_reachable_with(
+            mir,
+            &mut NeedlessCowVisitor {
+                cx,
+                cow_locals_to_assignments: &mut cow_locals_to_assignments,
+            },
+        );
+
+        for assignments in cow_locals_to_assignments.values() {
+            if assignments.len() == 1 {
+                continue;
+            }
+            let Some(first) = assignments.first() else {
+                continue;
+            };
+
+            span_lint_hir_and_then(
+                cx,
+                NEEDLESS_COW,
+                mir.source_scopes[first.source_info().scope]
+                    .local_data
+                    .as_ref()
+                    .assert_crate_local()
+                    .lint_root,
+                assignments.iter().map(|a| a.source_info().span).collect_vec(),
+                "usage of `Cow` where the `Borrowed` variant is only ever used for empty strings",
+                |diag| {
+                    diag.help("remove the `Cow` and use `String::new` instead, as it allocates no memory");
+
+                    for into_or_from in assignments.iter().filter(|a| matches!(a, CowAssignment::IntoOrFrom(_))) {
+                        diag.span_note(into_or_from.source_info().span, "this implicitly constructs `Borrowed`");
+                    }
+                },
+            );
+        }
+    }
+}
+
+type CowAssignments = Vec<CowAssignment>;
+
+#[derive(Clone, Copy)]
+enum CowAssignment {
```

Comment:
To properly handle `async` correctly, we need another variant that's for closures. If this closure isn't used in a larger `if`/`match`/etc., we would handle it as if it's its own separate "thing", but if it's part of a larger structure, we'll handle assignments within it as if there's no closure at all. Basically:

```rust
if true { (|| Cow::Borrowed(""))() } else { Cow::Borrowed("") }
```

Should lint, but:

```rust
(|| Cow::Borrowed(""))();
```

Should not, since the closure isn't declared and called within an `if`/`match`/etc.

Also, if the closure contains its own `if` expression, we'd handle that too.